src/code42cli/cmds/devices.py

```diff
@@ -0,0 +1,383 @@
+            except Py42NotFoundError:
+                full_disk_access_status = False
+        else:
+            full_disk_access_status = ""
+        current_result_dict["full_disk_access"] = full_disk_access_status
+        return current_result_dict
+
+    result_list = run_bulk_process(
+        handle_row, rows, progress_label="Getting device settings"
+    )
+    try:
+        return device_dataframe.merge(DataFrame.from_records(result_list), on="guid")
+    except KeyError:
+        return device_dataframe
+
+
+def _drop_devices_which_have_not_connected_in_some_number_of_days(
+    devices_dataframe, days_since_last_connected
+):
+    return devices_dataframe.loc[
+        to_datetime(datetime.now(), utc=True)
+        - to_datetime(devices_dataframe["lastConnected"], utc=True)
+        > to_timedelta(days_since_last_connected, unit="days"),
+        :,
+    ]
+
+
+def _drop_n_devices_per_user(
+    device_dataframe,
+    number_to_drop,
+    sort_field="lastConnected",
+    sort_ascending=False,
+    group_field="userUid",
+):
+    return (
+        device_dataframe.sort_values(by=sort_field, ascending=sort_ascending)
+        .drop(device_dataframe.groupby(group_field).head(number_to_drop).index)
+        .reset_index(drop=True)
+    )
+
+
+def _add_usernames_to_device_dataframe(sdk, device_dataframe):
+    users_generator = sdk.users.get_all()
+    users_list = []
+    for page in users_generator:
+        users_list.extend(page["users"])
+    users_dataframe = DataFrame.from_records(
+        users_list, columns=["username", "userUid"]
+    )
+    return device_dataframe.merge(users_dataframe, how="left", on="userUid")
+
+
+@devices.group(cls=OrderedGroup)
+@sdk_options(hidden=True)
+def bulk(state):
+    """Tools for managing devices in bulk."""
+    pass
+
+
+@bulk.command(
+    name="deactivate",
+    help="""Deactivate all devices on the given list.
+            Takes as input a CSV with a deviceId column.
+            The deviceId column must have the legacy deviceId value, not GUID.""",
+)
+@read_csv_arg(headers=["deviceId"])
+@change_device_name_option
+@purge_date_option
+@format_option
+@sdk_options()
+def bulk_deactivate(state, csv_rows, change_device_name, purge_date, format):
+    sdk = state.sdk
+    csv_rows[0]["deactivated"] = False
+    formatter = OutputFormatter(format, {key: key for key in csv_rows[0].keys()})
+    for row in csv_rows:
+        row["change_device_name"] = change_device_name
+        row["purge_date"] = purge_date
+
+    def handle_row(**row):
+        try:
+            _deactivate_device(
+                sdk, row["deviceId"], row["change_device_name"], row["purge_date"]
+            )
+            row["deactivated"] = "True"
+        except Exception as e:
+            row["deactivated"] = "False: {}".format(e)
+        return row
+
+    result_rows = run_bulk_process(
+        handle_row, csv_rows, progress_label="Deactivating devices:"
+    )
+    formatter.echo_formatted_list(result_rows)
```

Comment:
Cool, I think this is the first bulk command that outputs data after the process completes.. Is there any other intended purpose for this besides giving the user feedback? I am curious how this looks next to the progress bar and stats

Reply:
We could potentially add an option to silence the progress bar so the data is pipeable, if it needs to be?

Reply:
I think it already plays nice with piping as-is. I just tested with piping into a file, and the file only had the end results, not the progress bar.